protoc_plugin/test/message_generator_test.dart

```diff
@@ -87,4 +94,45 @@ void main() {
     expectMatchesGoldenFile(writer.sourceLocationInfo.toString(),
         'test/goldens/messageGeneratorEnums.meta');
   });
+
+  test('testMetadataIndices', () {
+    var options =
+        parseGenerationOptions(CodeGeneratorRequest(), CodeGeneratorResponse());
+    FileGenerator fg = FileGenerator(fd, options);
+    MessageGenerator mg =
+        MessageGenerator.topLevel(md, fg, {}, null, Set<String>(), 0);
+
+    var ctx = GenerationContext(options);
+    mg.register(ctx);
+    mg.resolve(ctx);
+
+    var writer = IndentingWriter(filename: '');
+    mg.generate(writer);
+
+    var eq = ListEquality();
+    var fieldStringsMap = HashMap(
+        equals: eq.equals, hashCode: eq.hash, isValidKey: eq.isValidKey);
+    fieldStringsMap[[4, 0]] = ['PhoneNumber'];
+    fieldStringsMap[[4, 0, 2, 0]] = ['type', 'hasType', 'clearType'];
+    fieldStringsMap[[4, 0, 2, 1]] = ['number', 'hasNumber', 'clearNumber'];
+    fieldStringsMap[[4, 0, 2, 2]] = ['name', 'hasName', 'clearName'];
+    fieldStringsMap[[4, 0, 2, 3]] = [
+      'deprecatedField',
+      'hasDeprecatedField',
+      'clearDeprecatedField'
+    ];
+
+    String generatedContents = writer.toString();
+    GeneratedCodeInfo metadata = writer.sourceLocationInfo;
+    for (var annotation in metadata.annotation) {
+      String annotatedName =
+          generatedContents.substring(annotation.begin, annotation.end);
+      var expectedStrings = fieldStringsMap[annotation.path];
+      if (expectedStrings == null) {
+        fail('The field path ${annotation.path} '
+            'did not match any expected field path.');
+      }
+      expect(annotatedName, isIn(expectedStrings));
```

Comment:
I like this explicit testing of the source code.

Reply:
Me too, I think this is a solid pattern.